[dagster-gcp] PipesDataprocJobClient tutorial (#27423)
dagster-io/dagster: docs/docs-beta/docs/guides/build/external-pipelines/gcp-dataproc-pipeline.md

---
title: Build pipelines with GCP Dataproc
description: "Learn to integrate Dagster Pipes with GCP Dataproc to launch external code from Dagster assets."
sidebar_position: 300
---

> This article covers how to use [Dagster Pipes](/guides/build/external-pipelines/) with [GCP Dataproc](https://cloud.google.com/dataproc) in Job mode.

Reviewer: I feel like "Job mode" could also be a link to relevant Dataproc docs

The [dagster-gcp](/api/python-api/libraries/dagster-gcp) integration library provides the <PyObject section="libraries" object="pipes.PipesDataprocJobClient" module="dagster_gcp" /> resource, which can be used to launch GCP Dataproc jobs from Dagster assets and ops. Dagster can receive regular events such as logs, asset checks, or asset materializations from jobs launched with this client. Using it requires minimal code changes to your Dataproc jobs.

> 

Reviewer: "regular events" -> "events"

("regular" sounds like it's on a schedule)

Reviewer: "Using it requires" -> "The client requires"

<details>
  <summary>Prerequisites</summary>

    - **In the Dagster environment**, you'll need to:

    - Install the following packages:

        ```shell
        pip install dagster dagster-webserver 'dagster-gcp[dataproc]'
        ```

        Refer to the [Dagster installation guide](/getting-started/installation) for more info.

    - **Configure GCP authentication for applications**. If you don't have this set up already, refer to the [GCP authentication guide](https://cloud.google.com/docs/authentication/gcloud).

    - **In GCP**, you'll need:

>     - An existing project with a Dataproc cluster.

Reviewer: Should be nested under previous bullet right?

    - Prepared infrastructure such as GCS buckets, IAM roles, and other resources required for your Dataproc job.

</details>

## Step 1: Install the dagster-pipes module in your Dataproc environment

Choose one of the [options](https://cloud.google.com/dataproc/docs/tutorials/python-configuration) to install `dagster-pipes` in the Dataproc environment.

For example, use the following property in your configuration:

```yaml
dataproc:pip.packages: "dagster-pipes,google-cloud-storage"
```

> `google-cloud-storage` is an optional dependency required for passing Pipes messages from the Dataproc job to dagster.

Reviewer 2: ```suggestion
`google-cloud-storage` is an optional dependency required for passing Pipes messages from the Dataproc job to Dagster.
```

Author: The CI was yelling at me for having `Dagster` here - that's why I changed it to lowercase... :melting_face:

## Step 2: Add dagster-pipes to the Dataproc job script

> Call `open_dagster_pipes` in the Dataproc script to create a context that can be used to send messages to dagster:

Reviewer 2: ```suggestion
Call `open_dagster_pipes` in the Dataproc script to create a context that can be used to send messages to Dagster:
```

<CodeExample path="docs_snippets/docs_snippets/guides/dagster/dagster_pipes/gcp/dataproc_job/script.py" />

:::tip

The metadata format shown above (`{"raw_value": value, "type": type}`) is part of Dagster Pipes' special syntax for specifying rich Dagster metadata. For a complete reference of all supported metadata types and their formats, see the [Dagster Pipes metadata reference](using-dagster-pipes/reference#passing-rich-metadata-to-dagster).

:::

## Step 3: Create an asset using the PipesDataprocJobClient to launch the job

In the Dagster asset/op code, use the `PipesDataprocJobClient` resource to launch the job:

<CodeExample path="docs_snippets/docs_snippets/guides/dagster/dagster_pipes/gcp/dataproc_job/dagster_code.py" startAfter="start_asset_marker" endBefore="end_asset_marker" />

> This will launch the Dataproc job and wait for it completion. If the job fails, the Dagster process will raise an exception. If the Dagster process is interrupted while the job is still running, the job will be terminated.

Reviewer 2: ```suggestion
This will launch the Dataproc job and wait for it to complete. If the job fails, the Dagster process will raise an exception. If the Dagster process is interrupted while the job is still running, the job will be terminated.
```
